File: Algorithms/0986_Interval_List_Intersections/Python/Interval_List_Intersections_Solution_1.py

# Space: O(1)
# Time: O(n)


class Solution:
    def intervalIntersection(self, firstList, secondList):
        if not firstList or not secondList: return []
        i = j = 0
        res = []

        while i < len(firstList) and j < len(secondList):
            left = max(firstList[i][0], secondList[j][0])
            right = min(firstList[i][1], secondList[j][1])
            if left <= right: res.append([left, right])
            if firstList[i][1] < secondList[j][1]:
                i += 1
            else:
                j += 1
        return res

File: Algorithms/0986_Interval_List_Intersections/Java/IntervalListIntersectionsSolution1.java

// Space: O(1)
// Time: O(n)

class Solution {
    public int[][] intervalIntersection(int[][] firstList, int[][] secondList) {
        List<int[]> res = new ArrayList<>();
        int i = 0, j = 0;
        int left = 0, right = 0;
        while (i < firstList.length && j < secondList.length) {
            left = Math.max(firstList[i][0], secondList[j][0]);
            right = Math.min(firstList[i][1], secondList[j][1]);
            if (left <= right) res.add(new int[]{left, right});

            if (firstList[i][1] < secondList[j][1]) {
                i++;
            } else {
                j++;
            }
        }
        return res.toArray(new int[res.size()][]);
    }
}




